policy.go

```diff
@@ -0,0 +1,309 @@
+		Size:       0}
+	index, err := tpm.NVDefineSpace(tpm.OwnerHandleContext(), nil, public, session)
+	if err != nil {
+		if isNVIndexDefinedError(err) {
+			return &nvIndexDefinedError{public.Index}
+		}
+		return xerrors.Errorf("cannot create NV index: %w", err)
+	}
+
+	succeeded := false
+	defer func() {
+		if succeeded {
+			return
+		}
+		tpm.NVUndefineSpace(tpm.OwnerHandleContext(), index, session)
+	}()
+
+	// Begin a session to initialize the index.
+	policySession, err := tpm.StartAuthSession(nil, nil, tpm2.SessionTypePolicy, nil, nameAlg)
+	if err != nil {
+		return xerrors.Errorf("cannot begin policy session to initialize NV index: %w", err)
+	}
+	defer tpm.FlushContext(policySession)
+
+	// Compute a digest for signing with our key
+	signDigest := tpm2.HashAlgorithmSHA256
+	h := signDigest.NewHash()
+	h.Write(policySession.NonceTPM())
+	binary.Write(h, binary.BigEndian, int32(0))
+
+	// Sign the digest
+	sig, err := rsa.SignPSS(rand.Reader, key, signDigest.GetHash(), h.Sum(nil), &rsa.PSSOptions{SaltLength: rsa.PSSSaltLengthEqualsHash})
+	if err != nil {
+		return xerrors.Errorf("cannot provide signature for initializing NV index: %w", err)
+	}
+
+	// Load the public part of the key in to the TPM. There's no integrity protection for this command as if it's altered in
+	// transit then either the signature verification fails or the policy digest will not match the one associated with the NV
+	// index.
+	keyLoaded, err := tpm.LoadExternal(nil, keyPublic, tpm2.HandleEndorsement)
+	if err != nil {
+		return xerrors.Errorf("cannot load public part of key used to initialize NV index to the TPM: %w", err)
+	}
+	defer tpm.FlushContext(keyLoaded)
+
+	signature := tpm2.Signature{
+		SigAlg: tpm2.SigSchemeAlgRSAPSS,
+		Signature: tpm2.SignatureU{
+			Data: &tpm2.SignatureRSAPSS{
+				Hash: signDigest,
+				Sig:  tpm2.PublicKeyRSA(sig)}}}
+
+	// Execute the policy assertions
+	if err := tpm.PolicyCommandCode(policySession, tpm2.CommandNVWrite); err != nil {
+		return xerrors.Errorf("cannot execute assertion to initialize NV index: %w", err)
+	}
+	if err := tpm.PolicyCounterTimer(policySession, clockBytes, 8, tpm2.OpUnsignedLT); err != nil {
+		return xerrors.Errorf("cannot execute assertion to initialize NV index: %w", err)
+	}
+	if _, _, err := tpm.PolicySigned(keyLoaded, policySession, true, nil, nil, 0, &signature); err != nil {
+		return xerrors.Errorf("cannot execute assertion to initialize NV index: %w", err)
+	}
+
+	// Initialize the index
+	if err := tpm.NVWrite(index, index, nil, 0, policySession, session.IncludeAttrs(tpm2.AttrAudit)); err != nil {
+		return xerrors.Errorf("cannot initialize NV index: %w", err)
+	}
+
+	// Marshal key name and cut-off time for writing to the NV index so that they can be used for verification in the future.
+	data, err := tpm2.MarshalToBytes(lockNVIndexVersion, keyName, time.ClockInfo.Clock)
+	if err != nil {
+		panic(fmt.Sprintf("cannot marshal contents for policy data NV index: %v", err))
```

Comment:
Could this be an errors as well? panic in the library code is a bit risky

Reply:
the target here is memory, the question is can there be untested bugs in MarshalToBytes that would trigger this vs runtime serious issues?  in any case if this is changed there is quite a lot of similar code in go-tpm2 as well that would need changing too